README.md
pyUspsLib
=========

NOTES
-------------------------
Requires Requests
USPS requires you run tests with their data before you can get production permissions. These tests are included.


Features
-------------------------
This package allows you to get data from USPS. It currently supports address verification, shipping pricing, and package tracking.


Installation
-------------------------

EASY :

	pip install pyuspslib

from github :

	pip install https://github.com/luxnovalabs/pyUspsLib/zipball/master


Running the USPS test data:
	
	from pyuspslib import tests
	user_id = 'my_usps_user_id'
	tests.run_all_tests(user_id)

	shipping domestic and international shipping pricing will fail.
	Despite the docuementation, you don't have access yet


After running the tests, give USPS a call. The instructions say to email, but I gave up after waiting 72 hours. The phone call took minutes.


BASIC USAGE
==================

	from pyuspslib.lib import USPS
	user_id = 'my_usps_user_id'

	usps = USPS(user_id)


all methods return a dictionary with "value", "error", and "xml" keys


	#find city and state for a given zipcode
	response = usps.loopkup_city_state('84115')
	print response['value'] # returns {'city':'value', 'state':'value'}
	print response['error'] # returns string from usps describing the error
	print response['xml'] # raw xml from python's xml.etree.ElementTree


	#returns a zipcode for the address information given
	response = usps.lookup_zipcode({'Address2':'244 Edison', 'City':'Salt Lake City', 'State':'UT'})
	print response['value'] # returns 84111
	print response['error'] # returns string from usps describing the error
	print response['xml'] # raw xml from python's xml.etree.ElementTree


	#returns a zipcode for the address information given
	response = usps.verify_address({'Address2':'244 Edison', 'City':'Salt Lake City', 'State':'UT'})
	print response['value'] # returns 'valid' or None
	print response['error'] # returns string from usps describing the error
	print response['xml'] # raw xml from python's xml.etree.ElementTree


	#returns status of a packge id
	response = usps.track_package('EJ958083578US')
	print response['value'] # returns 'valid' or None
	print response['error'] # returns string from usps describing the error
	print response['xml'] # raw xml from python's xml.etree.ElementTree



INTERMEDIATE USAGE
==================

USPS has specific fields that must be sent. With shipping prices, the structure gets a little complicated.

1) Start by copying the defaults.py file and creating a usps instance like this:

	from pyuspslib.lib import USPS
	import my_custom_defaults

	user_id = 'my_usps_user_id'

	usps = USPS(user_id, my_custom_defaults)

2) Take a look at the defaults file and change the origin zipcode

	USPS_MERCHANT_ORIGIN_ZIPCODE = 'your zipcode'

3) You should also set useful defaults for package defaults. Weight, size, and shape all affect the price. International pacakges need additonal information.

4) The methods provided are written to be simple, and pull the extra data needed from the defaults. If you need more control, see the advanced section.

	#returns price of domestic shipping
	#needs shipping_type and destination zipcode
	response = usps.price_domestic_shipping('PRIORITY','84111') 
	print response['value'] # returns a decimal price
	print response['error'] # returns string from usps describing the error
	print response['xml'] # raw xml from python's xml.etree.ElementTree


International shipping is more complicated. You're basically forced into advanced usage - which is still really simple.

USPS requires fields, even if empty. We'll pull the template from the defaults and stuff our values in the template.	

	from pyuspslib import defaults 

	#merge default data with new value
	package_data = defaults.USPS_DEFAULT_INTERNATIONAL_PACKAGE_DATA.update({'ValueOfContents':'50'})
	
	#returns a list of shipping options
	response = usps.price_international_shipping(package_data)
	print response['value'] # returns a set of shipping options
	print response['error'] # returns string from usps describing the error
	print response['xml'] # raw xml from python's xml.etree.ElementTree




ADVANCED USAGE
==================

The USPS library is written to provide some convenience, but also allow you to override everything. You've probably noticed that all methods return a dictionary with a 'xml' key. You can crawl that data as much as you like to get anything usps provides.

The methods, at first glance, allow limited data entry, however they have arguments with defaults you can override.

For example, if you are getting a domestic shipping estimate and the shipping type and zipcode are not enough, look at the methods  in lib.USPS to see what you can override.

Instead of this:
	
	response = usps.price_domestic_shipping('PRIORITY','84111')

you can call this:
	
	response usps.price_domestic_shipping(shipping_type, destination_zipcode, package_data=None, api='RateV4')

If package_data = None, the method pulls in the defaults.USPS_DEFAULT_DOMESTIC_PACKAGE_DATA and uses that. For custom data, bring in your own instance and override the values.
	
	package_data = defaults.USPS_DEFAULT_INTERNATIONAL_PACKAGE_DATA.update({'ValueOfContents':'50'})

and pass your package_data into the method



If the data structure changes on the USPS end, you can pass custom xml templates into the USPS constructor.





CHANGES.txt
v<0.1.0>, <08-04-2013> -- Initial release.
v<0.1.1>, <08-08-2013> -- Modifications to  shipping options.

pyuspslib/tests.py
import requests
import defaults

class address_test():

	def __init__(self, user_id, base_url):
		self.request = None
		self.user_id = user_id
		self.base_url = base_url

	def usps_address_verify(self, user_id, base_url, api='Verify'):
		url = '%s?API=%s&XML=<AddressValidateRequest USERID="%s">\
			<Address>\
				<Address1></Address1>\
				<Address2>6406 Ivy Lane</Address2>\
				<City>Greenbelt</City>\
				<State>MD</State> \
				<Zip5></Zip5>\
				<Zip4></Zip4>\
			</Address>\
		</AddressValidateRequest>' % (base_url, api, user_id)
		self.request = requests.get(url)

	def usps_zipcode_lookup(self, user_id, base_url, api='ZipCodeLookup'):
		url = '%s?API=%s&XML=<ZipCodeLookupRequest USERID="%s">\
			<Address>\
				<Address1></Address1>\
				<Address2>6406 Ivy Lane</Address2>\
				<City>Greenbelt</City>\
				<State>MD</State> \
			</Address>\
		</ZipCodeLookupRequest>' % (base_url, api, user_id)
		self.request = requests.get(url)


	def usps_city_state_lookup(self, user_id, base_url, api='CityStateLookup'):
		url = '%s?API=%s&XML=<CityStateLookupRequest USERID="%s">\
				<ZipCode ID= "0">\
					<Zip5>90210</Zip5>\
				</ZipCode>\
		</CityStateLookupRequest>' % (base_url, api, user_id)
		self.request = requests.get(url)

	def run_tests(self):
		self.usps_address_verify(self.user_id, self.base_url)
		print 'ADDRESS_VERIFY : %s' % self.request.text
		self.usps_zipcode_lookup(self.user_id, self.base_url)
		print 'ZIPCODE_LOOKUP : %s' % self.request.text
		self.usps_city_state_lookup(self.user_id, self.base_url)
		print 'CITY_STATE_LOOKUP : %s' % self.request.text


class tracking_test():

	def __init__(self, user_id, base_url):
		self.request = None
		self.user_id = user_id
		self.base_url = base_url

	def tracking_test(self, user_id, base_url, api='TrackV2'):
		url = '%s?API=%s&XML=<TrackRequest USERID="%s">\
			<TrackID ID="EJ958083578US"></TrackID>\
		</TrackRequest>' % (base_url, api, user_id)
		self.request = requests.get(url)

	def run_tests(self):
		self.tracking_test(self.user_id, self.base_url)
		print 'TACKING_TEST : %s' % self.request.text


class shipping_test():

	def __init__(self, user_id, base_url):
		self.request = None
		self.base_url = base_url
		self.user_id = user_id

	def domestic_shipping_test(self, user_id, base_url, api='RateV4'):
		url = '%s?API=%s&XML=<RateV4Request USERID="%s" >\
			<Revision/>\
				<Package ID="1ST">\
					<Service>PRIORITY</Service>\
					<ZipOrigination>44106</ZipOrigination>\
					<ZipDestination>20770</ZipDestination>\
					<Pounds>1</Pounds>\
					<Ounces>8</Ounces>\
					<Container>NONRECTANGULAR</Container>\
					<Size>LARGE</Size>\
					<Width>15</Width>\
					<Length>30</Length>\
					<Height>15</Height>\
					<Girth>55</Girth>\
				</Package>\
			</RateV4Request>' % (base_url, api, user_id)
		self.request = requests.get(url)

	def inernational_shipping_test(self, user_id, base_url, api='IntlRateV2'):
		url = '%s?API=%s&XML=<IntlRateV2Request USERID="%s">\
			<Revision>2</Revision> \
				<Package>\
					<Pounds>69</Pounds>\
					<Ounces>0</Ounces> \
					<Machinable>True</Machinable>\
					<MailType>all</MailType>\
					<GXG>\
						<POBoxFlag>N</POBoxFlag>\
						<GiftFlag>N</GiftFlag>\
					</GXG>\
					<ValueOfContents>100.00</ValueOfContents>\
					<Country>canada</Country>\
					<Container>RECTANGULAR</Container>\
					<Size>Regular</Size>\
					<Width>10</Width>\
					<Length>10</Length>\
					<Height>11</Height>\
					<Girth></Girth>\
					<CommercialFlag>y</CommercialFlag>\
				</Package>\
			</IntlRateV2Request>' % (base_url, api, user_id)
		self.request = requests.get(url)

	def run_tests(self):
		self.domestic_shipping_test(self.user_id, self.base_url)
		print 'SHIPPING_TEST : %s' % self.request.text
		self.inernational_shipping_test(self.user_id, self.base_url)
		print 'INTERNATIONAL_TEST : %s' % self.request.text



def run_all_tests(user_id):

	address = address_test(user_id, defaults.USPS_TEST_URL)
	address.run_tests()

	tracking = tracking_test(user_id, defaults.USPS_TEST_URL)
	tracking.run_tests()

	shipping = shipping_test(user_id, defaults.USPS_TEST_URL)
	shipping.run_tests()











pyuspslib/templates.py
domestic_shipping_template = '''<RateV4Request>
<Revision></Revision>
<Package ID="1ST">
<Service></Service>
<ZipOrigination></ZipOrigination>
<ZipDestination></ZipDestination>
<Pounds></Pounds>
<Ounces></Ounces>
<Container></Container>
<Size></Size>
<Width></Width>
<Length></Length>
<Height></Height>
<Girth></Girth>
<Machinable></Machinable>
</Package>
</RateV4Request>'''

international_shipping_template = '''<IntlRateV2Request>
<Revision>2</Revision>
<Package ID="1ST">
<Pounds></Pounds>
<Ounces></Ounces>
<Machinable></Machinable>
<MailType></MailType>
<GXG>
<POBoxFlag>N</POBoxFlag>
<GiftFlag>N</GiftFlag>
</GXG>
<ValueOfContents></ValueOfContents>
<Country></Country>
<Container></Container>
<Size></Size>
<Width></Width>
<Length></Length>
<Height></Height>
<Girth></Girth>
<CommercialFlag></CommercialFlag>
</Package>
</IntlRateV2Request>'''

address_template = '''<AddressValidateRequest>
<Address>
<Address1></Address1>
<Address2></Address2>
<City></City>
<State></State>
<Zip5></Zip5>
<Zip4></Zip4>
</Address>
</AddressValidateRequest>'''

zipcode_template = '''<ZipCodeLookupRequest>
<Address>
<Address1></Address1>
<Address2></Address2>
<City></City>
<State></State>
</Address>
</ZipCodeLookupRequest>'''

city_state_template = '''<CityStateLookupRequest>
<ZipCode ID="0">
<Zip5></Zip5>
</ZipCode>
</CityStateLookupRequest>'''

tracking_template = '''<TrackRequest>
<TrackID></TrackID>
</TrackRequest>'''


pyuspslib/lib.py
import requests
import xml.etree.ElementTree as XML
from decimal import Decimal

import defaults, templates, tests


class USPS():

	def __init__(self, user_id=None, settings_file=defaults, templates=templates):
		self.request = None
		self.user_id = user_id if user_id else settings_file.USPS_USER_ID
		self.base_shipping_fee = settings_file.USPS_BASE_SHIPPING_FEE
		self.merchant_zipcode = settings_file.USPS_MERCHANT_ORIGIN_ZIPCODE
		self.base_url = settings_file.USPS_TEST_URL if settings_file.USPS_TESTMODE else settings_file.USPS_PRODUCTION_URL
		#self.default_domestic_package_data = settings_file.USPS_DEFAULT_DOMESTIC_PACKAGE_DATA
		#self.default_international_package_data = settings_file.USPS_DEFAULT_INTERNATIONAL_PACKAGE_DATA
		self.default_domestic_package_data = {}
		self.templates = templates

	##########################
	#  ADDRESS VERIFICATION  #
	##########################
	def verify_address(self, address_data={}, api='Verify'):
		#login info
		xml = XML.fromstring(self.templates.address_template)
		xml.set('USERID', self.user_id)
		#address bindings
		address = xml.find('Address')
		for key, value in address_data.iteritems():
			address.find(key).text = value
		#request
		self.request = requests.get('%s?API=%s&XML=%s' % (self.base_url, api, XML.tostring(xml)))
		response_xml = XML.fromstring(self.request.text)
		#error checking
		error = response_xml.find('.//Description')
		response = {
			'error':error.text if error is not None else None,
			'value':True if error is None else False,
			'xml':response_xml
		}
		return response

	def lookup_zipcode(self, address_data={}, api='ZipCodeLookup'):
		#login info
		xml = XML.fromstring(self.templates.zipcode_template)
		xml.set('USERID', self.user_id)
		#address bindings
		address = xml.find('Address')
		for key, value in address_data.iteritems():
			address.find(key).text = value
		#request
		self.request = requests.get('%s?API=%s&XML=%s' % (self.base_url, api, XML.tostring(xml)))
		response_xml = XML.fromstring(self.request.text)
		#error checking
		error = response_xml.find('.//Description')
		response = {
			'error':error.text if error is not None else None,
			'value':response_xml.find('.//Zip5').text if error is None else error.text,
			'xml':response_xml
		}
		return response
		

	def loopkup_city_state(self, zipcode, api='CityStateLookup'):
		#login info
		xml = XML.fromstring(self.templates.city_state_template)
		xml.set('USERID', self.user_id)
		#address bindings
		xml.find('.//Zip5').text = zipcode
		#request
		self.request = requests.get('%s?API=%s&XML=%s' % (self.base_url, api, XML.tostring(xml)))
		response_xml = XML.fromstring(self.request.text)
		#error checking
		error = response_xml.find('.//Description')
		city = response_xml.find('.//City').text if error is None else None
		state = response_xml.find('.//State').text if error is None else None
		response = {
			'error':error.text if error is not None else None,
			'value':{'city':city, 'state':state},
			'xml':response_xml
		}
		return response

	##########################
	#  	  SHIPPING PRICES    #
	##########################
	#dimensions.lbs, .ounces, .container_shape, .size, .width, .length, .height, .girth
	def price_domestic_shipping(self, shipping_type, destination_zipcode, package_data=None, api='RateV4'):
		details = package_data if package_data else self.default_domestic_package_data
		#login info
		xml = XML.fromstring(self.templates.domestic_shipping_template)
		xml.set('USERID', self.user_id)
		#package binding
		package = xml.find('Package')
		package.find('ZipOrigination').text = self.merchant_zipcode
		package.find('ZipDestination').text = destination_zipcode
		package.find('Service').text = shipping_type
		for key, value in details.iteritems():
			package.find(key).text = value
		#request
		self.request = requests.get('%s?API=%s&XML=%s' % (self.base_url, api, XML.tostring(xml)))
		response_xml = XML.fromstring(self.request.text)
		#error checking
		error = response_xml.find('.//Description')
		rate = Decimal(response_xml.find('.//Rate').text) if error is None else None
		response = {
			'error':error.text if error is not None else None,
			'value':rate,
			'xml':response_xml
		}
		return response

		
	def price_international_shipping(self, package_data, flags={}, api='IntlRateV2'):
		details = package_data if package_data else self.default_international_package_data
		#login info
		xml = XML.fromstring(self.templates.international_shipping_template)
		xml.set('USERID', self.user_id)
		#package data
		package = xml.find('Package')
		for key, value in details.iteritems():
			package.find(key).text = value
		#gifts and POBox
		gxg = package.find('GXG')
		for key, value in flags.iteritems():
			gxg.find(key).text = value
		#request
		self.request = requests.get('%s?API=%s&XML=%s' % (self.base_url, api, XML.tostring(xml)))
		response_xml = XML.fromstring(self.request.text)
		error = response_xml.find('.//Description')
		#returns multiple shipping options
		services = response_xml.find('Package').findall('Service')
		rates={}
		if error is None:
			for service in services:
				service_id = service.get('ID')
				Postage = service.find('Postage')
				CommercialPostage = service.find('CommercialPostage')
				SvcCommitments = service.find('SvcCommitments')
				SvcDescription = service.find('SvcDescription')
				rates[str(service_id)] = {
					'Postage': Decimal(Postage.text) if Postage is not None else None,
					'CommercialPostage':Decimal(CommercialPostage.text) if CommercialPostage is not None else None,
					'SvcCommitments':SvcCommitments.text if SvcCommitments is not None else None,
					'SvcDescription':SvcDescription.text if SvcDescription is not None else None,
				}
		response = {
			'error':error.text if error is not None else None,
			'value': rates,
			'xml':response_xml
		}
		return response
		


	###############
	#  TRACKING   #
	###############
	def track_package(self, tracking_id, api='TrackV2'):
		#login info
		xml = XML.fromstring(self.templates.tracking_template)
		xml.set('USERID', self.user_id)
		#tacking_id
		tracking_element = xml.find('TrackID')
		tracking_element.set('ID', tracking_id)
		#request
		self.request = requests.get('%s?API=%s&XML=%s' % (self.base_url, api, XML.tostring(xml)))
		response_xml = XML.fromstring(self.request.text)
		#error checking
		error = response_xml.find('.//Description')
		status = response_xml.find('.//TrackSummary')
		response = {
			'error':error.text if error is not None else None,
			'value': status.text if status is not None else None,
			'xml':response_xml
		}
		return response

######################
#  UTILITIES  #
######################
# def step_dictionary(xml, obj):
# 	for key, value in obj.iteritems():
# 		child = XML.SubElement(xml, key)
# 		if type(value) is dict:
# 			step_dictionary(child, value)
# 		else:
# 			child.text = value


# def dictionary_to_xml(obj, user_id, root_element):
# 	root = XML.Element(root_element)
# 	root.set('USERID', user_id)
# 	step_dictionary(root, obj)
# 	return root


def ounces_to_lbs(ounces):
	#returns object with lbs and ounces
	pass

def lbs_to_ounces(lbs):
	pass


#step 1 run all tests and then call usps and get production certified
#tests.run_all_tests(defaults.USPS_USER_ID)

#step 2 run these examples check the results
# response = usps.loopkup_city_state('84115')
# print response['status']

# response = usps.lookup_zipcode({'Address2':'244 Edison', 'City':'Salt Lake City', 'State':'UT'})
# print response['status']

# response = usps.verify_address({'Address2':'244 Edison', 'City':'Salt Lake City', 'State':'UT'})
# print response['status']

# response = usps.track_package('EJ958083578US')
# print response['status']

# response = usps.price_domestic_shipping('PRIORITY','84105')
# print response['rate']

# package_data = defaults.USPS_DEFAULT_INTERNATIONAL_PACKAGE_DATA.update({'ValueOfContents':'50'})
# response = usps.price_international_shipping(package_data)
# print len(response['services'])
